### ISSUE TITLE
Rule proposal: No variable named after special name

### ISSUE THREAD

**Issue by jfmengels** [sindresorhus/eslint-plugin-unicorn#42]
In my codebase at work, I often see variables/params named as `module` which is kind of dangerous in my opinion.

I'd like to propose a rule that forbids the use of "reserved" variables, or at least the "special" variables, such as: `module`, `exports`, `global`, `window` (?), ...

I'd also like to see the same thing but for node core modules, like `fs`, `path`, that would forbid the following for instance

``` js
var path = '/foo/bar';
```

Not sure if we'd want that in a new rule or behind a flag. I'm leaning towards the latter.

**Comment by sindresorhus** [comment]
:+1:

**Comment by jfmengels** [comment]
I found a list of variables to be protected for Node here https://nodejs.org/api/globals.html. Don't know enough about front-end dev to protect anything but the same ones + `window`.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/globals

**Comment by jfmengels** [comment]
Should have guessed it :smile:

**Comment by sindresorhus** [comment]
Actually, the first part is already covered by the [`no-global-assign`](http://eslint.org/docs/rules/no-global-assign) rule:

```js
module = true;
```

```js
index.js:1:1
  ✖  1:1   Read-only global module should not be modified.  no-global-assign
```

-

But I think preventing using the same name as built-in Node.js modules is a good idea too. Suggestion for rule name? `no-nodejs-builtin-assign`?

**Comment by jfmengels** [comment]
> Actually, the first part is already covered by the no-global-assign rule:

In part, yes, but not really, as it wouldn't save us from the following case, which is what I see in our codebase.

```js
function foo(module) {
    // Do something with `module`
}
```

> But I think preventing using the same name as built-in Node.js modules is a good idea too. Suggestion for rule name? no-nodejs-builtin-assign?

`no-nodejs-builtin-shadow`? To reference [`no-shadow`](http://eslint.org/docs/rules/no-shadow)

**Comment by sindresorhus** [comment]
> In part, yes, but not really, as it wouldn't save us from the following case, which is what I see in our codebase.

Ah, ok. So still value in having a rule for this then.

> o-nodejs-builtin-shadow? To reference no-shadow

I would go with `no-shadow-nodejs-builtins` to match [`no-shadow-restricted-names`](http://eslint.org/docs/rules/no-shadow-restricted-names).

**Comment by florianb** [comment]
Isn't this case somehow covered by eslints [`no-shadow`](http://eslint.org/docs/rules/no-shadow) (with `builtinGlobals: true`) and [no-shadow-restricted-names](http://eslint.org/docs/rules/no-shadow-restricted-names) rules?

But i must confess i wasn't able to completely dig through how the globals are evaluated.

**Comment by fregante** [comment]
> Isn't this case somehow covered by eslints [`no-shadow`](http://eslint.org/docs/rules/no-shadow) (with `builtinGlobals: true`

It is:

### No globals

```ts
var path = '/foo/bar';
var exports = 'wow';
var window = 'wow';
```

```
  3:5  error  'exports' is already declared in the upper scope  no-shadow
  4:5  error  'window' is already declared in the upper scope   no-shadow

✖ 2 problems (2 errors, 0 warnings)
```

But `path` isn't covered, only globals are. 

<details><summary>RunKit code</summary>

Copy to https://npm.runkit.com/eslint

```js
var eslint = require("eslint")

const code = `
var path = '/foo/bar';
var exports = 'wow';
var window = 'wow';
`;

const config = {
    "parserOptions": {
        "ecmaVersion": 5,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "no-shadow": [
            "error",
            {"builtinGlobals": true}
        ]
    },
    "env": {
        "browser": true,
        "node": true
    }
};

const { ESLint } = require("eslint");

(async function main() {
  const eslint = new ESLint({ overrideConfig: config,useEslintrc: false });
  const results = await eslint.lintText(code);
  const formatter = await eslint.loadFormatter("stylish");
  const resultText = formatter.format(results);
  console.log(resultText);
})().catch(console.error);
```

</details>


### No reserved words

ESLint comes with its own [id-denylist](https://eslint.org/docs/rules/id-denylist), but you'd have to configure it with a list of built-in modules… while still allowing the import of such modules.
